### ISSUE TITLE
Split package problem when using java modules

### ISSUE THREAD

**Issue by hendrikebbers** [langchain4j/langchain4j#1066]
langchain4j can not be used in a Java project that is working on the module path. By doing so the following error happens:

`error: the unnamed module reads package dev.langchain4j.retriever from both langchain4j and langchain4j.core
`
That is based on having the same package `dev.langchain4j.retriever` (with content) in both modules `langchain4j` and `langchain4j.core`.

![image](https://github.com/langchain4j/langchain4j/assets/9443847/0337d42e-d0cf-4643-86ce-5b70d42e51e8)

A project like langchain4j should be 100% compatible with the Java module system since that is used by more and more libs over time.

I did a lot of work in that area and happy to help if you want to work on that topic.

**Comment by koppor** [comment]
In our project, we used the [hints by Nicolai Parlog](https://nipafx.dev/five-command-line-options-hack-java-module-system/#adding-classes-to-modules-with---patch-module) to have a workaround:

```diff
+        '--add-exports=langchain4j/dev.langchain4j.model.chat=org.jabref',
+        '--add-exports=langchain4j/dev.langchain4j.model.chat=org.jabref.merged.module',
...
+modularity.patchModule("langchain4j", "langchain4j-core-0.28.0.jar")
+modularity.patchModule("langchain4j", "langchain4j-embeddings-0.28.0.jar")
+modularity.patchModule("langchain4j", "langchain4j-embeddings-all-minilm-l6-v2-0.28.0.jar")
```

**Comment by langchain4j** [comment]
Hi @hendrikebbers, thanks a lot for reporting!

Since `dev.langchain4j.retriever.Retriever` is deprecated since 0.26, we could get rid of it in the next (0.31) release.
Do you see any other way around it?
Thanks!

**Comment by koppor** [comment]
There are many more issues with "split pacakges":

![image](https://github.com/langchain4j/langchain4j/assets/1366654/5fb03660-d3f0-4906-8dcb-08de359fd780)

**Comment by koppor** [comment]
- `dev.langchain4j.agent.tool from both langchain4j and langchain4j.core`
- `dev.langchain4j.classification from both langchain4j and langchain4j.core`
- `dev.langchain4j.chain from both langchain4j and langchain4j.core`
- `dev.langchain4j.code from both langchain4j and langchain4j.core`
- `dev.langchain4j.model.embedding from both langchain4j.embeddings.all.minilm.l6.v2 and langchain4j.embeddings`
- `dev.langchain4j.model.output from both langchain4j and langchain4j.core`

**Comment by langchain4j** [comment]
Oh :(

**Comment by hendrikebbers** [comment]
I assume we have two problems here: 

- `langchain4j` and `langchain4j.core` have a lot of packages in common. Here it would be great to introduce a custom package for one of them. Example: all content of `langchain4j` will be under the package `dev.langchain4.impl.*` in future. By doing so you would have the interface `dev.langchain4.impl.retriever.Retriever` in `langchain4j` and the class`dev.langchain4.retriever.EmbeddingStoreRetriever` in `langchain4j.core`. Is my assumption (api in `langchain4j.core` and implementations in`langchain4j`) correct?
- `langchain4j.embeddings` shares packages with core (`dev.langchain4j.model.embedding`). Maybe all content in the embeddings jars can be moved to `dev.langchain4j.embedding.*`?

**Comment by konczdev** [comment]
This is also a problem in osgi environment. The preferred way to wire bundles are import/export packages but this is not working with split packages. The solution is require bundle in the manifest, but it is tighter coupling. I think either langchain and langchain-core should be merged or the package names should be changed. As I see the most recent issues, there will be breaking changes in the near future, so maybe this is the time when it is possible to done.

**Comment by hendrikebbers** [comment]
@langchain4j if you want / need help in that refactoring or just for review please ping me. Happy to help

**Comment by langchain4j** [comment]
Hi @hendrikebbers, `langchain4j-core` now has APIs and some implementaitons useful for many use cases. I was considering extracting API into `langchain4j-api` or something like that. So it might look like this: `langchain4j` -> `langchain4j-core` -> `langchain4j-api`.
I need a bit more time to analze where breaking changes are acceptable and where it would be nicer to avoid it.

**Comment by langchain4j** [comment]
@konczdev could you please elaborate on breaking changes in the near future? We usually try to avoid breaking changes as much as possible.

**Comment by konczdev** [comment]
A few examples: #1042 #1044 #1048 #1049 #1050

**Comment by konczdev** [comment]
> Hi @hendrikebbers, `langchain4j-core` now has APIs and some implementaitons useful for many use cases. I was considering extracting API into `langchain4j-api` or something like that. So it might look like this: `langchain4j` -> `langchain4j-core` -> `langchain4j-api`. I need a bit more time to analze where breaking changes are acceptable and where it would be nicer to avoid it.

I think it's already confusing what the role of the langchain4j and langchain4j-core modules are. Whatever wil be the solution, each module should have its own package namespace and the api and implementation modules should not share the same package.

**Comment by langchain4j** [comment]
@konczdev #1042 and #1044 are not supposed to be breaking changes. #1048 and #1049 are, but this functionality is not widely used AFAIK, so it is OKish. #1050 is not decided yet how to proceed.

But I agree that there should be no split-packages. Let's see how to fix this problem with as less breaking changes as possible.

**Comment by konczdev** [comment]
@langchain4j For me those are breaking changes too because I have to wrap non-OSGi dependencies into OSGi bundles. Maybe after the resolution of this issue I will make some contribution to generate the proper OSGi manifests at compile time.

**Comment by konczdev** [comment]
@langchain4j Don't get me wrong, I didn't mean to criticize anything :) langchain4j is a wonderful project, I follow its development day by day.

**Comment by langchain4j** [comment]
@konczdev all good, constructive criticism is always welcome! And you are providing very valuable feedback, so thank you! I will have a bit more time next week to look into this problem more closely, so we can continue this discussion then, ok?

My only concern is to avoid introducing unnecessary breaking changes as much as possible, but of course some breaking changes will **_have_** to be done sooner or later, so let's do that sooner. I just need a bit more time to analyze it.

**Comment by hendrikebbers** [comment]
Sounds reasonable! It is better to have a set of breaking changes in 1 release than to have new breaking changes in every release.

**Comment by koppor** [comment]
The embeddig modles also seems re-use packages: Modules `langchain4j.embeddings.all.minilm.l6.v2.q` and `langchain4j.embeddings.all.minilm.l6.v2` export package `dev.langchain4j.model.embedding` to module jersey.client

**Comment by koppor** [comment]
This is kind of blocking, because one cannot compile using IntelliJ anymore: `--patch-module`  requires a full JAR path, but this is different for each developer.

Current state at our side:

![image](https://github.com/langchain4j/langchain4j/assets/1366654/96037777-7069-4689-828f-299aded8dec4)

**Comment by langchain4j** [comment]
@koppor going to address this asap in the next release

**Comment by koppor** [comment]
@langchain4j Thank you! Then, I will suspend work on workarounds.  (Smalltalk: Currently, we can neither run using gradle on Windows (https://github.com/java9-modularity/gradle-modules-plugin/issues/281) nor execute in IntelliJ at all. Thus, we need to develop on Linux or macOS.)

Looking foward! If there is a SNAPSHOT to test, please let us know, we are really happy to test!

**Comment by hendrikebbers** [comment]
again happy to help. @langchain4j if you want we can do a meeting the next days to discuss how that could be solved.

**Comment by langchain4j** [comment]
Hi @hendrikebbers, thank you for offering your help!
AFAIU the only way to fix this is to either rename conflicting packages or merge/move classes from same packages into one artifact (e.g. move some classes from `langchain4j` into `langchain4j-core`), right?
So I guess I will have to analyze each case separately one by one to find the least breaking solution.

Or do you have another plan/idea?
Thanks!

**Comment by konczdev** [comment]
Packages should have a unique origin, so no two modules must contain types in the same package.

Just moving some classes won't resolve the issue. Langchain4j.module.a and langchain4j.module.b cannot share the same langchain4j.module.somepackage package name.

**Comment by langchain4j** [comment]
@konczdev, here is an example of what I mean by moving/merging:

Before:
```
module langchain4j-core:
    package dev.langchain4j.chain:
        class Chain

module langchain4j:
    package dev.langchain4j.chain:
        class ConversationalChain
        class ConversationalRetrievalChain
```

After:
```
module langchain4j-core:
    package dev.langchain4j.chain:
        class Chain
        class ConversationalChain
        class ConversationalRetrievalChain

module langchain4j:
    (no package dev.langchain4j.chain)
```

Moving `ConversationalChain` and `ConversationalRetrievalChain` from `langchain4j` to `langchain4j-core` should resolve the issue, no?

**Comment by konczdev** [comment]
@langchain4j in that case yes. I would also change the base package to dev.langchain4j.core which would help you to avoid accidental package name conflict in the future. (sorry for the missing formatting, written from mobile)

**Comment by langchain4j** [comment]
@konczdev good idea, but let me check the impact first

**Comment by hendrikebbers** [comment]
agree with @konczdev - It would be best if each module (jar) has a custom base package at same level. Having 1 jar with `dev.langchain4j` as base and a second with `dev.langchain4j.core` can end in adding `core` in the first one by accident.

**Comment by hendrikebbers** [comment]
Maybe it makes sense to split this and start with everything that is not core first.

Multiple non-core modules (like `langchain4j-cassandra`, `langchain4j-azure-ai-search`, `langchain4j-neo4j`) have similar issues. It could be defined that extension must have `dev.langchain4j.store.embedding.FOO` or `dev.langchain4j.model.FOO` as base package.

Better would even to have `dev.langchain4j.extension.FOO` (could be `dev.langchain4j.extension.FOO.store.embedding` or `dev.langchain4j.extension.FOO.model` than) to have a real hard split. But that would end in changing all modules.

**Comment by langchain4j** [comment]
@hendrikebbers I am trying to keep changes to the minimum to not break everything.
I see your point with each module having it's own package prefix, but this would mean big-bang changes for every user.

Is there a way to have an automatic check for split packages in the CI? I can resolve current split-package issues without much breaking changes and once CI check is in place, we are safe.

**Comment by langchain4j** [comment]
I've drafted a fix for all curent split package issues, please take a look:
- https://github.com/langchain4j/langchain4j/pull/1433
- https://github.com/langchain4j/langchain4j-embeddings/pull/33

**Comment by konczdev** [comment]
Looking only at the PRs it looks good, but I don't have a local environment currently. So @koppor @hendrikebbers, can you check it too?

**Comment by langchain4j** [comment]
Fixes are available in `0.33.0-SNAPSHOT` version, please check it.
Don't forget to alow snapshot dependencies:
```xml
    <repositories>
        <repository>
            <id>snapshot-repo</id>
            <url>https://s01.oss.sonatype.org/content/repositories/snapshots</url>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
        </repository>
    </repositories>
```

**Comment by hendrikebbers** [comment]
I will have a look. First look at PRs looks good

**Comment by hendrikebbers** [comment]
@langchain4j  To do full support for the module system, we need to add at least basic names to each module. While for full support, a `module-info.java` would be needed, the easiest way is to add the info to the `MANIFEST` of the jar as an `automatic-module-name`. You can find more information about that here https://dev.java/learn/modules/automatic-module/
Based on that, I created a branch and defined the names of each module. You can find my changes in this draft: https://github.com/langchain4j/langchain4j/pull/1502

Sadly I can not run `mvn install` on my machine to verify that everything works for a project that uses modules since the build on my machine always ends with the following errors. I did the same for main branch to verify that it is not my change that did it. All errors happen for me on mainbranch, too.

```
[INFO] --- compiler:3.12.1:testCompile (default-testCompile) @ langchain4j ---
[INFO] Recompiling the module because of changed dependency.
[INFO] Compiling 37 source files with javac [debug target 8] to target/test-classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithRagIT.java:[12,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/store/embedding/inmemory/InMemoryEmbeddingStoreSerializedTest.java:[4,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/classification/EmbeddingModelTextClassifierTest.java:[3,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/store/embedding/inmemory/InMemoryEmbeddingStoreRemovalTest.java:[4,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/store/embedding/inmemory/InMemoryEmbeddingStoreTest.java:[6,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[INFO] 5 errors 
```

If I ignore test code and do a `mvn compile` I can go some steps further but end with the following errors:

```
[INFO] --- compiler:3.12.1:compile (default-compile) @ langchain4j-easy-rag ---
[INFO] Recompiling the module because of changed dependency.
[INFO] Compiling 1 source file with javac [debug target 8] to target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j-easy-rag/src/main/java/dev/langchain4j/data/document/splitter/recursive/RecursiveDocumentSplitterFactory.java:[5,39] cannot find symbol
  symbol:   class HuggingFaceTokenizer
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j-easy-rag/src/main/java/dev/langchain4j/data/document/splitter/recursive/RecursiveDocumentSplitterFactory.java:[12,57] cannot find symbol
  symbol:   class HuggingFaceTokenizer
  location: class dev.langchain4j.data.document.splitter.recursive.RecursiveDocumentSplitterFactory
[INFO] 2 errors 
```

**Comment by hendrikebbers** [comment]
Created discussion https://github.com/langchain4j/langchain4j/discussions/1503 to not hide the problem (that others might have) in this issue.

**Comment by langchain4j** [comment]
@hendrikebbers thanks a lot for the review and your PR, I will check it asap!

**Comment by InAnYan** [comment]
@langchain4j Hi! Just bumped langchain version to 0.33. Works perfectly without any workaround :smile: ! Thanks!

**Comment by chraledev** [comment]
With the 1.0.0-alpha1 the problem ist back.

```
langchain-core: dev.langchain4j.exception.UnsupportedFeatureException
langchain4j: dev.langchain4j.exception.IllegalConfigurationException
```

```
Error occurred during initialization of boot layer
java.lang.module.ResolutionException: Module langchain4j.core contains package dev.langchain4j.exception, module langchain4j exports package dev.langchain4j.exception to langchain4j.core
```

Proposition: move both exceptions to langchain4j-core

**Comment by dliubarskyi** [comment]
@chraledev thanks a lot for the heads up! Fixed in https://github.com/langchain4j/langchain4j/pull/2466

I have also created https://github.com/langchain4j/langchain4j/issues/2467 to automathe the check for split packages.

**Comment by konczdev** [comment]
> agree with @konczdev - It would be best if each module (jar) has a custom base package at same level. Having 1 jar with `dev.langchain4j` as base and a second with `dev.langchain4j.core` can end in adding `core` in the first one by accident. 

I know this would be a huge breaking change.
